src/viam/ur/module/ur_arm_state.cpp

```diff
@@ -130,30 +131,38 @@ std::unique_ptr<URArm::state_> URArm::state_::create(std::string configured_mode
                                           std::move(csv_output_path),
                                           std::move(threshold),
                                           std::move(frequency),
-
                                           ports);
 
     state->set_speed(degrees_to_radians(find_config_attribute<double>(config, "speed_degs_per_sec").value()));
     state->set_acceleration(degrees_to_radians(find_config_attribute<double>(config, "acceleration_degs_per_sec2").value()));
 
+    // Hold the mutex while we start the worker thread. It will not be
+    // able to advance, but will be ready to take over work as soon as
+    // we release the lock, minimizing the delay between establishing
+    // a connection and allowing the worker thread to begin meeting
+    // its service obligations.
     const std::lock_guard lock(state->mutex_);
     state->worker_thread_ = std::thread{&state_::run_, state.get()};
 
     // Attempt to manually drive the state machine out of the
     // disconnected state. For now, just try a few times with the
     // natural recovery cycle.
-    for (size_t i = 0; i != 5; ++i) {
+    size_t connection_failures = 0;
```

Comment:
This loop needed a refactor because calls to `upgrade_downgrade_` now may be either initiating a connection or polling for completion, so just calling it 5 times doesn't work anymore. Instead, we just call it repeatedly until we either get 5 failures or a success. Effectively, that's the same observable behavior.